File: src/main/java/net/dontdrinkandroot/gitki/wicket/page/SignInPage.java
package net.dontdrinkandroot.gitki.wicket.page;

import com.giffing.wicket.spring.boot.context.scan.WicketSignInPage;
import net.dontdrinkandroot.gitki.service.user.UserService;
import net.dontdrinkandroot.gitki.wicket.GitkiWebSession;
import net.dontdrinkandroot.gitki.wicket.headeritem.StyleCssHeaderItem;
import org.apache.wicket.RestartResponseAtInterceptPageException;
import org.apache.wicket.markup.head.HeaderItem;
import org.apache.wicket.model.IModel;
import org.apache.wicket.model.Model;
import org.apache.wicket.request.mapper.parameter.PageParameters;
import org.apache.wicket.spring.injection.annot.SpringBean;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
@WicketSignInPage
public class SignInPage extends net.dontdrinkandroot.wicket.bootstrap.page.SignInPage
{
    @SpringBean
    private UserService userService;

    public SignInPage(PageParameters parameters)
    {
        super(parameters);
    }

    @Override
    protected void onInitialize()
    {
        if (!this.userService.hasAdminUser()) {
            throw new RestartResponseAtInterceptPageException(FirstRunPage.class);
        }

        super.onInitialize();
    }

    @Override
    protected IModel<String> createUsernameLabelModel()
    {
        return Model.of("Email");
    }

    @Override
    protected boolean isSignedIn()
    {
        return null != GitkiWebSession.get().getUser();
    }

    @Override
    protected boolean signIn(String username, String password)
    {
        return GitkiWebSession.get().signIn(username, password);
    }

    @Override
    protected HeaderItem getBootstrapCssHeaderItem()
    {
        return new StyleCssHeaderItem();
    }
}

File: src/main/kotlin/net/dontdrinkandroot/gitki/wicket/page/SignInPage.kt
package net.dontdrinkandroot.gitki.wicket.page

import com.giffing.wicket.spring.boot.context.scan.WicketSignInPage
import net.dontdrinkandroot.gitki.service.user.UserService
import net.dontdrinkandroot.gitki.wicket.getGitkiSession
import net.dontdrinkandroot.gitki.wicket.head.AppCssHeaderItem
import net.dontdrinkandroot.wicket.bootstrap.page.SignInPage
import org.apache.wicket.RestartResponseAtInterceptPageException
import org.apache.wicket.markup.head.HeaderItem
import org.apache.wicket.model.IModel
import org.apache.wicket.model.Model
import org.apache.wicket.request.mapper.parameter.PageParameters
import org.apache.wicket.spring.injection.annot.SpringBean

@WicketSignInPage
open class SignInPage(parameters: PageParameters) : SignInPage(parameters) {

    @SpringBean
    private lateinit var userService: UserService

    override fun onInitialize() {
        assertAdminUserExists()
        super.onInitialize()
    }

    private fun assertAdminUserExists() {
        if (!userService.hasAdminUser()) throw RestartResponseAtInterceptPageException(FirstRunPage::class.java)
    }

    override fun createUsernameLabelModel(): IModel<String> = Model.of("Email")

    override val signedIn: Boolean
        get() = null != getGitkiSession().user

    override fun signIn(username: String, password: String): Boolean = getGitkiSession()
        .signIn(username, password)

    override val bootstrapHeaderItem: HeaderItem
        get() = AppCssHeaderItem()
}